[bee #48] pushsync: initial pushsync implementation

acud opened the pull request:
Naive implementation.

Defer to later stage:
- [ ] create a simple recorder that allows for an isolated protocol instance to be tested, then cleanup this code to use that recorder

closes #48

pradovic commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+	for {
+		select {
+		// handle incoming chunks
+		case ch, more := <-chunks:
```
Sorry if I am asking the stupid question, I am not very familiar with the algorithm here.  When is `chunks` channel being written to? 🙈

pradovic commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+		Storer:        storer,
+	})
+	defer ps.Close()
+	recorder.SetProtocols(ps.Protocol())
```
Why is protocol being sent on the same recorder? Sorry, I might misunderstand something cuz I am not familiar with the protocol 🙈

acud replied:
recorder in its current form is not suitable for this kind of testing that expects a protocol to initiate the creation of its own streams, that is, there is a mutual dependency between recorder and the protocol instance.

the pr description mentions that a simpler recorder implementation is needed, since recorder in its current form actually needs two protocol instances, but this is not needed for this test vector, since there is no initiating "other" node which creates the stream, only one unit is under test here. if you have any other suggestion i am open to hear it out.

janos replied:
I agree, but, since we will have receipts, I think that the two sided recorder can be used fully in these tests and maybe shorten them a bit, compared to explicitly testing handler or client function.

pradovic replied:
That's what I meant. Basically, we created streamtest very opinionated, to test both sides and not just one side. If you want to test one side you could basically just mock read, write and create stream calls, and no need to use streamtest at all. Do you think guys it would make sense to create mocks for new stream, read and write calls, and then use them if we want to test just one side, either handler or sender? It will be also easier when we abstract creating streams in general

pradovic replied:
Sorry, maybe it's not the scope of this PR just a general thought :)

pradovic commented:
@acud Nice work! I left a few comments, might be because I don't understand some of fine detailis of the protocol.

acud commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+
+	// setup the stream recorder to record stream data
+	recorder := streamtest.New(
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
```
please read the code. it is needed so that `handler` will not be called

nolash replied:
shouldn't we have just `streamtest.WithNoHandler` that conceals exactly this?

Why wouldn't I call `streamtest.New().WithMiddleWares()`. I thought it was convention now to use chainables?

pradovic replied:
No need to get defensive, I was just curious as it was not obvious to me from the code or the comment ✌️

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+		case 1:
+			break LOOP
+		case 0:
+			time.Sleep(time.Millisecond * 10)
```
ewww

janos replied:
Is total of 100ms to wait good enough for the slow ci?

acud replied:
i will raise it

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(messages) > 1 {
```
can this happen when you already checked `Records()` above?

acud replied:
well, i guess anything can happen. do you want to leave it to fate? or to comparators?

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+		t.Fatal(err)
+	}
+
+	for i := 0; i < 20; i++ {
```
does the divine inspiration for these numbers come from your levitation meditation?

acud replied:
...

jmozah replied:
No. my third eye said so

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if l := len(records); l != 1 {
```
why not use the same construction as in the previous test here for enforcing only one record?

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+				// windows has a bit lower time resolution
+				// so, slow down the handler with a middleware
+				// not to get 0s for rtt value
+				time.Sleep(100 * time.Millisecond)
```
`ewww^2`

janos replied:
I think that this is from the pingpong test and do not think that it is needed here.

nolash commented on pkg/pushsync/pushsync_test.go:
```diff
@@ -2,4 +2,336 @@
+	})
+	defer ps.Close()
+
+	// TODO: this needs to go away once we have a simpler recorder
```
repeating this comment doesn't make it go away more

nolash commented:
Oh, I forgot to mention there is a conspicuous lack of vogon commentary in the code

jmozah commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+	Logger logging.Logger
+}
+
+var retryInterval = 10 * time.Second // time interval between retries
```
like do you want to have a flag during startup to indicate timeout?

nolash replied:
Sorry, I should have been clearer, I was thinking of the numeric value.

jmozah replied:
it can be.. but for now this naive implementation i set it as constant

jmozah commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+	}
+
+	// push this to the closest node too
+	peer, err := ps.peerSuggester.ClosestPeer(addr)
```
forward nowhere or to all nodes.. whats your point.?

nolash replied:
Hm never mind. The original push-sync implementation sought to send to neighborhood, but I guess it's not the case here, and wouldn't even have enough information to do so.

acud commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+		case ch, more := <-chunks:
+			// if no more, set to nil, reset timer to 0 to finalise batch immediately
+			if !more {
+				chunks = nil
```
L134

nolash replied:
Ah, I see, sneaky.

janos commented on pkg/pushsync/export_test.go:
```diff
@@ -0,0 +1,7 @@
+package pushsync
```
Add copyright header.

janos commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+	}
+	defer streamer.Close()
+
+	w, _ := protobuf.NewWriterAndReader(streamer)
```
This also come to my attention, but in the next PR we should have receipts, and the reader will be needed. It is a minor thingT he most correct thing would be to use NewWriter just as in handler NewReader should be used. But I do not think that the change is necessary.

jmozah replied:
This function just sends the message... so you are right.. i will change to Writer

acud commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,218 @@
+			// retry interval timer triggers starting from new
+		case <-timer.C:
+			// initially timer is set to go off as well as every time we hit the end of push index
+			func() {
```
i think it was originally done because probably some `defer` call to measure time... removing this

acud commented:
@pradovic I have addressed your comments about mocking the topology driver, it has reduced code in the tests, thanks.
@nolash I also addressed your comments about reusing the functionality to wait for a certain amount of records in the recorder and have just extracted that method to be on `Recorder`, to wait there directly, or fail the test when the requirement is not satisfied.

acud commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,216 @@
+		quit:          make(chan struct{}),
+	}
+
+	ctx := context.Background()
```
@janos WDYT about getting rid of `quit` in favor of keeping a `quitFunc`, that will just call the cancel function of this context?

acud replied:
but then, should the context not be cancelled when the quit is called? since this context is being passed to other components. cancelling it would make sure any pending operations are cancelled

janos replied:
In this case, there is no cancel function there is only a background context.

But this is a good point. This context is not doing any cancelation in this code, it is just shared without much purpose. I do not think that it is needed at all, and chunksWorker can create its own contexts where a function with a context should be called.

acud commented on pkg/pushsync/pushsync.go:
```diff
@@ -3,3 +3,215 @@
+			}
+
+			// retry interval timer triggers starting from new
+		case <-timer.C:
```
also, another thought is that this ugliness could be removed by making `SubscribePush` to just iterate until its empty. I think that this timer does not necessarily improve anything, but just slows things down. We have no proof that this actually makes things better. I think that infinitely iterating on push index will be nicer, rate limiting can be done implicitly by the length of the time it takes for the handler here to handle each chunk that is sent (i.e. if it is a blocking or an asynchronous operation)
cc @janos

janos replied:
I agree that this is a bit strange. Even if there are chunks in from `chunks` channel it is possible to get to this case when the timer fires and resets the subscription. The reason to retry would be if there are chunks put in the push index at earlier store time then the chunks that we receive from the subscription. I do not think that this could happen. I think that this is a good suggestion. 👍

acud replied:
> The reason to retry would be if there are chunks put in the push index at earlier store time then the chunks that we receive from the subscription

Then why shouldn't the subscription just iterate over all items in the index? this has to be done anyway to flush it empty no?

janos replied:
It does the iterate over all items, but iteration is only in forward.
